File: examples/ruby-rails/app/controllers/reference_controller.rb
require "json"
require "date"
require "net/http"
require "securerandom"
require "time"
require "timeout"
require_relative "../../lib/limited_body_reader"

class ReferenceController < ApplicationController
  MAXIMUM_BODY_BYTES = 64 * 1_024
  MAXIMUM_UPSTREAM_BYTES = 64 * 1_024
  TICKET_DEADLINE_SECONDS = 15
  SESSION_COOKIE = "cormier_session"
  RFC3339_TIMESTAMP = /\A\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})\z/

  def self.valid_future_expiration?(value, now = Time.now.utc)
    return false unless value.is_a?(String) && RFC3339_TIMESTAMP.match?(value)

    components = Date._iso8601(value)
    Date.valid_date?(components.fetch(:year), components.fetch(:mon), components.fetch(:mday)) &&
      Time.iso8601(value) > now
  rescue ArgumentError, KeyError
    false
  end

  def index
    shared_asset("index.html", "text/html; charset=utf-8")
  end

  def css
    shared_asset("app.css", "text/css; charset=utf-8")
  end

  def javascript
    shared_asset("app.js", "text/javascript; charset=utf-8")
  end

  def sdk_asset
    name = params[:asset].to_s
    return head :not_found unless /\A[A-Za-z0-9._-]+\z/.match?(name)

    send_asset(File.join(settings.sdk_asset_root, name), name.end_with?(".js") ? "text/javascript; charset=utf-8" : "application/json")
  end

  def health
    redis.ping
    render json: { status: "healthy" }
  rescue Redis::BaseError
    render json: { status: "unavailable" }, status: :service_unavailable
  end

  def diagnostics
    redis_status = begin
      redis.ping
      "ready"
    rescue Redis::BaseError
      "unavailable"
    end
    render json: {
      stack: "Ruby / Rails",
      topology: settings.topology,
      instance: settings.instance_name,
      redis: redis_status,
      timestamp: Time.now.utc.iso8601
    }
  end

  def login
    return unless require_origin

    tenant = params[:tenantId].to_s
    user = params[:userId].to_s
    unless settings.allows?(tenant, user)
      return render json: { code: "invalid_identity", message: "Select a configured test tenant and user." }, status: :bad_request
    end

    id = SecureRandom.urlsafe_base64(36, false)
    expires_at = Time.now.utc + settings.session_lifetime
    record = { tenantId: tenant, userId: user, allowedTopics: %w[orders notifications], expiresAt: expires_at.iso8601(6), revoked: false }
    redis.set(settings.session_key(id), JSON.generate(record), ex: settings.session_lifetime)
    cookies[SESSION_COOKIE] = {
      value: id,
      expires: expires_at,
      httponly: true,
      same_site: :strict,
      secure: settings.public_origin.start_with?("https://")
    }
    session[:reference_authenticated] = true
    render json: record
  rescue Redis::BaseError
    dependency_unavailable
  end

  def session_status
    record = read_session
    return render json: { code: "authentication_required", message: "Authentication is required." }, status: :unauthorized unless record

    render json: { authenticated: true }.merge(record)
  rescue Redis::BaseError, JSON::ParserError
    dependency_unavailable
  end

  def logout
    return unless require_origin

    id = cookies[SESSION_COOKIE].to_s
    redis.del(settings.session_key(id)) if valid_session_id?(id)
    reset_session
    cookies.delete(SESSION_COOKIE)
    head :no_content
  rescue Redis::BaseError
    dependency_unavailable
  end

  def ticket
    return unless require_origin
    return render(json: { code: "authentication_required", message: "Authentication is required." }, status: :unauthorized) unless read_session
    return render(json: { code: "invalid_request", message: "The request is invalid." }, status: :content_too_large) if request.content_length.to_i > MAXIMUM_BODY_BYTES

    body = LimitedBodyReader.read(request.body, limit: MAXIMUM_BODY_BYTES)

    status, content_type, response_body = forward_ticket(body)
    render body: response_body, status: status, content_type: content_type
  rescue LimitedBodyReader::TooLarge
    render json: { code: "invalid_request", message: "The request is invalid." }, status: :content_too_large
  rescue IOError, SystemCallError, Timeout::Error, Redis::BaseError, JSON::ParserError, LimitedBodyReader::ResponseTooLarge
    dependency_unavailable
  end

  private

  def shared_asset(name, type)
    send_asset(File.join(settings.shared_asset_root, name), type)
  end

  def send_asset(path, type)
    return head :not_found unless File.file?(path)

    send_file path, type: type, disposition: "inline"
  end

  def valid_session_id?(id)
    /\A[A-Za-z0-9_-]{16,256}\z/.match?(id)
  end

  def read_session
    return nil unless session[:reference_authenticated] == true

    id = cookies[SESSION_COOKIE].to_s
    return nil unless valid_session_id?(id)

    encoded = redis.get(settings.session_key(id))
    return nil unless encoded

    record = JSON.parse(encoded)
    expires_at = record.fetch("expiresAt")
    return nil unless self.class.valid_future_expiration?(expires_at)
    return nil if record.fetch("revoked", true)

    record
  rescue KeyError, ArgumentError
    nil
  end

  def forward_ticket(body)
    uri = settings.gateway_uri.dup
    uri.path = "/realtime/tickets"
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == "https"
    http.open_timeout = 5
    http.read_timeout = 15
    http.write_timeout = 5
    upstream = Net::HTTP::Post.new(uri)
    upstream["Host"] = request.host_with_port
    upstream["X-Forwarded-Proto"] = settings.public_uri.scheme
    %w[Origin Cookie Content-Type].each do |header|
      value = request.headers[header]
      upstream[header] = value if value.present?
    end
    upstream.body = body
    status = nil
    content_type = nil
    response_body = nil
    Timeout.timeout(TICKET_DEADLINE_SECONDS) do
      http.start do |connection|
        connection.request(upstream) do |response|
          status = response.code.to_i
          content_type = response["Content-Type"] || "application/json"
          response_body = LimitedBodyReader.collect(limit: MAXIMUM_UPSTREAM_BYTES) do |append|
            response.read_body { |chunk| append.call(chunk) }
          end
        end
      end
    end
    [ status, content_type, response_body ]
  end
end

File: examples/php-laravel/app/Http/Controllers/ReferenceController.php
<?php

namespace App\Http\Controllers;

use App\ReferenceConfig;
use Illuminate\Http\Client\ConnectionException;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Http\Response;
use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Redis;
use Illuminate\Support\Str;
use Throwable;

final class ReferenceController
{
    private const string COOKIE = 'cormier_session';

    private const int MAXIMUM_BODY_BYTES = 65536;

    public function __construct(private readonly ReferenceConfig $config) {}

    public function health(): JsonResponse
    {
        try {
            Redis::connection()->ping();

            return response()->json(['status' => 'healthy']);
        } catch (Throwable) {
            return response()->json(['status' => 'unavailable'], 503);
        }
    }

    public function diagnostics(): JsonResponse
    {
        try {
            Redis::connection()->ping();
            $redis = 'ready';
        } catch (Throwable) {
            $redis = 'unavailable';
        }

        return response()->json([
            'stack' => 'PHP / Laravel',
            'topology' => $this->config->topology,
            'instance' => $this->config->instanceName,
            'redis' => $redis,
            'heartbeatIntervalMilliseconds' => $this->config->heartbeatIntervalMilliseconds,
            'timestamp' => now('UTC')->toIso8601String(),
        ]);
    }

    public function login(Request $request): JsonResponse
    {
        if (($failure = $this->rejectOrigin($request)) !== null) {
            return $failure;
        }
        if ((int) $request->server('CONTENT_LENGTH', 0) > self::MAXIMUM_BODY_BYTES) {
            return $this->invalidIdentity();
        }
        $payload = $request->validate([
            'tenantId' => ['required', 'string', 'max:128'],
            'userId' => ['required', 'string', 'max:128'],
        ]);
        if (! $this->config->allows($payload['tenantId'], $payload['userId'])) {
            return $this->invalidIdentity();
        }
        $id = Str::random(48);
        $expiresAt = now('UTC')->addSeconds($this->config->sessionLifetimeSeconds);
        $record = [
            'tenantId' => $payload['tenantId'],
            'userId' => $payload['userId'],
            'allowedTopics' => ['orders', 'notifications'],
            'expiresAt' => $expiresAt->toIso8601String(),
            'revoked' => false,
        ];
        try {
            Redis::connection()->setex($this->config->sessionKey($id), $this->config->sessionLifetimeSeconds, json_encode($record, JSON_THROW_ON_ERROR));
        } catch (Throwable) {
            return $this->unavailable();
        }

        return response()->json($record)->cookie(
            self::COOKIE,
            $id,
            (int) ceil($this->config->sessionLifetimeSeconds / 60),
            '/',
            null,
            str_starts_with($this->config->publicOrigin, 'https://'),
            true,
            false,
            'Strict',
        );
    }

    public function session(Request $request): JsonResponse
    {
        try {
            $record = $this->readSession($request);
        } catch (Throwable) {
            return $this->unavailable();
        }
        if ($record === null) {
            return response()->json(['code' => 'authentication_required', 'message' => 'Authentication is required.'], 401);
        }

        return response()->json(['authenticated' => true] + $record);
    }

    public function logout(Request $request): Response|JsonResponse
    {
        if (($failure = $this->rejectOrigin($request)) !== null) {
            return $failure;
        }
        $id = (string) $request->cookie(self::COOKIE, '');
        if (preg_match('/^[A-Za-z0-9_-]{16,256}$/D', $id)) {
            try {
                Redis::connection()->del($this->config->sessionKey($id));
            } catch (Throwable) {
                return $this->unavailable();
            }
        }

        return response('', 204)->withoutCookie(self::COOKIE);
    }

    public function ticket(Request $request): Response|JsonResponse
    {
        if (($failure = $this->rejectOrigin($request)) !== null) {
            return $failure;
        }
        if (strlen($request->getContent()) > self::MAXIMUM_BODY_BYTES) {
            return response()->json(['code' => 'invalid_request', 'message' => 'The request is invalid.'], 413);
        }
        try {
            $upstream = Http::connectTimeout(5)->timeout(15)
                ->withOptions(['stream' => true])
                ->withHeaders(array_filter([
                    'Host' => $request->getHttpHost(),
                    'Origin' => $request->header('Origin'),
                    'Cookie' => $request->header('Cookie'),
                    'Content-Type' => $request->header('Content-Type'),
                    'X-Forwarded-Proto' => $this->config->publicScheme(),
                ]))
                ->withBody($request->getContent(), $request->header('Content-Type', 'application/json'))
                ->post($this->config->gatewayUrl.'/realtime/tickets');
            $stream = $upstream->toPsrResponse()->getBody();
            $body = '';
            while (! $stream->eof()) {
                $chunk = $stream->read(min(8192, self::MAXIMUM_BODY_BYTES + 1 - strlen($body)));
                if ($chunk === '' && ! $stream->eof()) {
                    throw new ConnectionException('The upstream response stream stalled.');
                }
                $body .= $chunk;
                if (strlen($body) > self::MAXIMUM_BODY_BYTES) {
                    return $this->unavailable();
                }
            }
        } catch (Throwable) {
            return $this->unavailable();
        }

        return response($body, $upstream->status())
            ->header('Content-Type', $upstream->header('Content-Type') ?: 'application/json')
            ->header('Cache-Control', 'no-store');
    }

    /** @return array<string, mixed>|null */
    private function readSession(Request $request): ?array
    {
        $id = (string) $request->cookie(self::COOKIE, '');
        if (! preg_match('/^[A-Za-z0-9_-]{16,256}$/D', $id)) {
            return null;
        }
        $encoded = Redis::connection()->get($this->config->sessionKey($id));
        if (! is_string($encoded)) {
            return null;
        }
        $record = json_decode($encoded, true, flags: JSON_THROW_ON_ERROR);
        if (! is_array($record) || ($record['revoked'] ?? true) !== false ||
            ! self::validFutureExpiration($record['expiresAt'] ?? null)) {
            return null;
        }

        return $record;
    }

    public static function validFutureExpiration(mixed $value, ?int $now = null): bool
    {
        if (! is_string($value) || ! preg_match(
            '/\A(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})(?:\.(\d{1,6}))?(Z|[+-]\d{2}:\d{2})\z/D',
            $value,
            $parts,
        )) {
            return false;
        }
        $fraction = $parts[2] ?? '';
        $zone = $parts[3] === 'Z' ? '+00:00' : $parts[3];
        $normalized = $parts[1].($fraction === '' ? '' : '.'.str_pad($fraction, 6, '0')).$zone;
        $format = $fraction === '' ? '!Y-m-d\TH:i:sP' : '!Y-m-d\TH:i:s.uP';
        $expiration = \DateTimeImmutable::createFromFormat($format, $normalized);
        $errors = \DateTimeImmutable::getLastErrors();
        if ($expiration === false || $errors !== false) {
            return false;
        }
        $microseconds = $expiration->format('u');
        $fractionMatches = $fraction === '' ||
            (str_starts_with($microseconds, $fraction) && trim(substr($microseconds, strlen($fraction)), '0') === '');

        return $expiration->format('Y-m-d\TH:i:s') === $parts[1] &&
            $expiration->format('P') === $zone && $fractionMatches && $expiration->getTimestamp() > ($now ?? time());
    }

    private function rejectOrigin(Request $request): ?JsonResponse
    {
        return $request->header('Origin') === $this->config->publicOrigin
            ? null
            : response()->json(['code' => 'origin_rejected', 'message' => 'The request Origin is not allowed.'], 403);
    }

    private function invalidIdentity(): JsonResponse
    {
        return response()->json(['code' => 'invalid_identity', 'message' => 'Select a configured test tenant and user.'], 400);
    }

    private function unavailable(): JsonResponse
    {
        return response()->json(['code' => 'service_unavailable', 'message' => 'The reference application dependency is unavailable.'], 503);
    }
}
